### ISSUE TITLE
[bug]This regular expression code will report an error

### ISSUE THREAD

**Issue by fenginsc** [scala-native/scala-native#3631]
A very simple regular expression that determines whether a string is a number or not compiles successfully with the latest version of scala-native, 0.4.16, but fails to run. 
```
object Main {
    def main(args: Array[String]): Unit = {
        println("start")
        if ("^(\\-|\\+)?(0\\.[0-9]+|[1-9][0-9]*\\.[0-9]+|[1-9][0-9]*|0)$".r.matches("1")) {
            println("OK")
        }
        println("end")
    }
}
```
The error is reported as follows:
```
scala.MatchError: 3 (of class java.lang.Integer)
        at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)
        at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)
        at scala.scalanative.unsafe.Zone$.open(Unknown Source)
        at java.lang.StackTrace$.currentStackTrace(Unknown Source)
        at java.lang.Throwable.printStackTrace(Unknown Source)
        at scala.scalanative.regex.Parser.removeLeadingRegexp(Unknown Source)
        at scala.scalanative.regex.Parser.factor(Unknown Source)
        at scala.scalanative.regex.Parser.collapse(Unknown Source)
        at scala.scalanative.regex.Parser.alternate(Unknown Source)
        at scala.scalanative.regex.Parser.parseRightParen(Unknown Source)
        at scala.scalanative.regex.Parser.scala$scalanative$regex$Parser$$parseInternal(Unknown Sour
ce)
        at scala.scalanative.regex.Parser$.parse(Unknown Source)
        at scala.scalanative.regex.RE2$.compileImpl(Unknown Source)
        at scala.scalanative.regex.Pattern$.compile(Unknown Source)
        at scala.scalanative.regex.Pattern$.compile(Unknown Source)
        at java.util.regex.Pattern$.compile(Unknown Source)
        at java.util.regex.Pattern$.LITERAL(Unknown Source)
        at java.util.regex.Pattern.compile(Unknown Source)
        at scala.collection.StringOps$.r$extension(Unknown Source)
        at Main$.main(Unknown Source)
        at Main.main(Unknown Source)
        at <none>.main(Unknown Source)
        at <none>.__scrt_common_main_seh:(D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_c
ommon.inl:288)(Unknown Source)
        at <none>.BaseThreadInitThunk(Unknown Source)
        at <none>.RtlUserThreadStart(Unknown Source)
```
I am using windows 10 x64, scala version is 2.13.12, sbt version is 1.9.8 and scala-native version is 0.4.16.

**Comment by LeeTibbert** [comment]
fenginsc,

Thank you for reporting this defect. Sorry that you are experiencing it.

The provided code made it easier to take a first look at this.

I confirm that the provided code works using scala-cli with Scala 3.3.1, & JVM (17) and
fails with Scala Native 4.16 (and probably others, including 5.0-SNAPSHOT).

The difference in my configuration is that I am using Linux (6.6.8).   

I suggest removing the "on windows" from the topic line now that the defect
has been reproduced on another OS.  I do not have the privs necessary to do that edit.

I am off to see if I can minimize the regex.

**Comment by ekrich** [comment]
I confirmed this works in Scala JVM as well.

**Comment by LeeTibbert** [comment]
Eric,  thank you for the data. OS?  I presume macOS.

**Comment by LeeTibbert** [comment]
Next Step:   I need to check the RE2 GitHub website to see if there are fixes there which
                     might resolve this Issue.

Progress:
* I have been able to isolate a simpler reproducer.
* The working goal is for SN to parse the presented Regex.
   I have been able to create a slightly altered Regex which works for both SN & JVM.
   The idea is to understand the problem by perturbing the inputs.
   `"^(\\-|\\+)?(0\\.[0-9]+)|([1-9][0-9]*\\.[0-9]+)|([1-9][0-9]*)|0$"

**Comment by LeeTibbert** [comment]
So that I understand, is the original regex just to illustrate the problem or is it intended to be
used in production?

If the latter, is the intent that  inputs have been put in some cannonical form so that then
regex need not recognize otherwise valid numbers such as "00" or "02.3"?

**Comment by fenginsc** [comment]
Thank you, I've changed the title, it was just an exercise to test if a string was a legitimate number, obviously "00" or "02.3" is not a mathematically legitimate number.

**Comment by LeeTibbert** [comment]
re: obviously "00" or "02.3" is not a mathematically legitimate number.

That is not so obvious to Java. Running the each of the lines below in scastie
succeeds and returns the expected value.

Granted that those strings do not represent the canonical or most reduced
form of a Platonic number.
 
```
java.lang.Integer.parseInt("00")
java.lang.Double.parseDouble("01.618")
```

**Comment by LeeTibbert** [comment]
@fenginsc 

At first blush, this defect looked like the defect reported fixed in PRs #2410 & #1701. They were merged into Scala Native
4.n & 5.n two plus years ago.

Further study narrowed the defect down to a mistake I made porting from the original re2j code five or so
years ago. I apologize for the time you have spent on my error  and the inconvenience it caused.

Thank you, once again, for flushing it out and reporting it.

PR #3642 contains a fix and corresponding test.  

I have marked that PR for backporting to the 4.n stream.

I do not know when the next 4.n release will be.  A release of Scala itself is coming up and 
Scala Native 4.n release which supports the new Scala is usually released very soon after.

Until then, can you use the workaround  of using parenthesis around any OR clause which
has more than two elements? In "(a+b+c+)| z" the first clause has three elements. Without
the parentheses, I believe that will provoke the reported defect. With them, I believe that
Scala Native will handle the string.  

If you can not use the workaround, let me know here and I will see what else I might be able
to figure out.

This one was a pretty good brain teaser.

**Comment by fenginsc** [comment]
@LeeTibbert 
Thank you very much for fixing this issue, I've fixed it by simplifying that regular expression `"^[-+]? (0|[1-9][0-9]*)(\\. [0-9]+)? $"` ran my program successfully, so maybe I won't be able to use the previous regular expression, it's just an example to demonstrate the bug. happy new year!
